maltoolbox/attackgraph/node.py

```diff
@@ -181,6 +183,101 @@ def is_available_defense(self) -> bool:
             'suppress' not in self.tags and \
             self.defense_status != 1.0
 
+    def ttc_sample(self) -> float:
+        """Sample a value from ttc distribution for a node
+
+        TTC Distributions in MAL:
+        https://mal-lang.org/mal-langspec/apidocs/org.mal_lang.langspec/org/mal_lang/langspec/ttc/TtcDistribution.html
+        """
+
+        def sample(exponential: float, bernoulli=1.0):
+            """
+            Generate a random sample for the given distributions.
+            If bernoulli distribution is not given, the sample will
+            simply be from an exponential.
+
+            If the Bernoulli trial fails (0), return infinity (impossible).
+            If the Bernoulli trial succeeds (1), return sample from
+            exponential distribution.
+            """
+
+            # If bernoulli is set to 1, the sample is just exponential
+            if np.random.choice([0, 1], p=[1 - bernoulli, bernoulli]):
+                return np.random.exponential(scale=1 / exponential)
+            return math.inf
+
+        if self.type == "defense":
+            # Defenses have no ttc
+            return 0
+
+        distribution = self.ttc.get('name')
+        s = math.nan
+        if distribution == "EasyAndCertain":
+            s = sample(exponential=1)
+        elif distribution == "EasyAndUncertain":
+            s = sample(exponential=1, bernoulli=0.5)
+        elif distribution == "HardAndCertain":
+            s = sample(exponential=0.1)
+        elif distribution == "HardAndUncertain":
+            s = sample(exponential=0.1, bernoulli=0.5)
+        elif distribution == "VeryHardAndCertain":
+            s = sample(exponential=0.01)
+        elif distribution == "VeryHardAndUncertain":
+            s = sample(exponential=0.01, bernoulli=0.5)
+        elif distribution == "Exponential":
+            scale = float(self.ttc['arguments'][0])
+            s = sample(exponential=scale)
+        else:
+            raise ValueError(f"Unknown TTC distribution: {distribution}")
+
+        return s
+
+    def ttc_expected_value(self) -> float:
+        """Returns the expected value of the ttc distribution for a node.
+
+        TTC Distributions in MAL:
+        https://mal-lang.org/mal-langspec/apidocs/org.mal_lang.langspec/org/mal_lang/langspec/ttc/TtcDistribution.html
+        """
```

Comment:
Why is this needed? Didn't we discuss in the last meeting that sampling TTC distributions would return the TTC value directly?